_posts/2019-10-11-hunting-down-a-dask-bug.markdown

```diff
@@ -0,0 +1,174 @@
+
+The API provides the data from a dataset (let's call it _output dataset_), which is
+filled from another dataset using Dask/Distributed (let's call this one _input dataset_)
+by shuffling around the data. The shuffling is done because the input dataset is
+organized in a way that reflects the customer's stores (basically one [Apache
+Parquet](https://parquet.apache.org) file per store), while the output dataset is
+organized by time (one Parquet file per day, containing data for all the stores).
+
+## The solution (as it seemed)
+
+A first investigation revealed that indeed, the store the customer complained about
+was not contained in the ouptut dataset, whereas it was contained in the input dataset.
+Therefore, the data must be lost somewhere in between during the shuffling.
+
+We had a look ad the Distributed cluster and sure enough, we discovered it was in a
+weird state. As it turned out, the day before, some of the hardware nodes hosting this
+cluster had been rebooted. This affected the scheduler and two of the workers, while
+the two other workers of the cluster were running on nodes that were not rebooted.
+As it looked like, after the reboot of the scheduler, these two workers were not able to
+connect to the new scheduler, so we ended up with a cluster with only two workers,
+while the two old workers were running in an endless loop trying to connect.
+
+Our assumption was that somehow, parts of the computations got scheduled to the invalid
+workers which did not deliver a result and thus the data got lost.
+To resolve this, we restarted the scheduler and all workers once again to get a
+healthy cluster. We re-ran the data shuffling process from the input dataset
+to the output dataset and checked that the store that had previously been missing was
+contained in the output dataset now. Thus, we informed the customer that the problem
+had been resolved.
+
+## Data loss strikes again
+
+Unfortunately, the customer got back to us, reporting that now a *different* store was
+missing. Indeed, we could verify that the problem persisted and each time, a different
+**random** store was missing. We also checked whether any changes had been done to the
+system recently, but besides the node reboots, nothing had been changed for weeks,
+in particular, we had been running exactly the same software versions.
+
+At this point in time, it became clear to us that we had to dig deeper into this.
+We began manually retracing each of the steps done during the data shuffling.
+The first step is reading in the data from the input dataset into a Dask distributed
+dataframe. A distributed dataframe can
+be seen as a large dataframe that is divided into several partitions that potentially
+live on different Distributed workers.
+We could verify that after this step, all of the stores were still contained
+in the data, so reading was not the problem. However, we made an interesting observation
+at this point. After reading in the data, we have one partition
```

Comment:
see above, "we had"

Reply:
Hm, we still do it that way, so the present should also be possible, right?